Build expressions

eklimo opened the pull request.

eklimo committed:
Build array expressions

eklimo committed:
Build expression statements

eklimo committed:
Build assignment expressions

eklimo committed:
Build literal expressions

eklimo committed:
Build cast expressions

eklimo committed:
Build operator expressions

eklimo committed:
Build `this`/`super` expressions

eklimo committed:
Build type reference expressions

eklimo committed:
Build field/variable expressions

eklimo committed:
Build method call expressions

eklimo committed:
Build ternary expressions

eklimo committed:
Refactor `AbstractApexNode`

- Add default implementations for some methods
- Add empty constructor

eklimo commented on pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/ast/ApexTreeBuilder.kt:
```diff
@@ -142,6 +177,174 @@ class ApexTreeBuilder(val sourceCode: String, val parserOptions: ApexParserOptio
+            BinaryExpression.Operator.RIGHT_SHIFT_UNSIGNED,
+            BinaryExpression.Operator.BITWISE_AND,
+            BinaryExpression.Operator.BITWISE_OR,
+            -> ASTBinaryExpression(node)
```
Should this be converted to a utility function in Summit?

aaronhurst-google replied:
The partition is just one of many possible ones. Dunno if it has uses outside of PMD.

As far as factoring to avoid complexity here... at least this looks fine to me.

eklimo commented on pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/ast/ApexTreeBuilder.kt:
```diff
@@ -142,6 +177,174 @@ class ApexTreeBuilder(val sourceCode: String, val parserOptions: ApexParserOptio
+    private fun flattenExpression(
+        node: Expression?,
+        components: List<Identifier> = emptyList()
+    ): Triple<List<Identifier>, Expression?, Boolean> =
```
Should the `Triple` be replaced with a `data class`?

zorzella replied:
My gut feeling is "probably yes", but I defer to you and Aaron.

aaronhurst-google replied:
It doesn't offend me... but Kotlin does make data classes so very easy. :-)

aaronhurst-google commented on pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/ast/AbstractApexNode.java:
```diff
@@ -107,15 +101,21 @@ public Iterable<? extends ApexNode<?>> children() {
     }
 
-    abstract void calculateLineNumbers(SourceCodePositioner positioner);
+    void calculateLineNumbers(SourceCodePositioner positioner) {
```
What do you think about leaving `AbstractApexNode` abstract and creating a `AbstractApexNode.Empty`?

At least that seems symmetrical with `AbstractApexNode.Single` and `AbstractApexNode.Many`. Are there any downsides?

eklimo replied:
All of these classes are `abstract`. Do you mean leaving the methods `abstract` as well and instead providing these default implementations in `Empty`?

aaronhurst-google replied:
Thanks, yes, that was what I intended to suggest. The reason being that if the derived class wants to inherit a set a common behaviors that is clear from the parent's name: `AbstractApexNode.Single`, `AbstractApexNode.Many`, `AbstractApexNode.Empty`. And then inheriting from just `AbstractApexNode` includes no assumptions about Summit<->PMD mapping, which will all be provided by the derived class.

eklimo replied:
I have been thinking about this, and although I initially agreed, I'm not sure if it's the right way to go in all cases. For a node like [`ASTEmptyReferenceExpression`](https://github.com/eklimo/pmd/pull/8/commits/9c23644b22a165cd9e447d1f35ef76a8bf272542#diff-278ba4e72958bc9750552cf0bde82002d8d0338d836456651148a913be8e2b39), which neither wraps a node directly (type parameter) nor indirectly (`Node` fields), the name `Empty` seems appropriate. However, a node like [`ASTField`](https://github.com/eklimo/pmd/pull/11/commits/57b76534b1d44b714bc52f9240f69d01262e69cc#diff-981dc32a5c105a05091f9b379c3c93914f6202b30b758b6c06e0182773e1f050) does indirectly wrap nodes through its fields, so the name `Empty` seems less appropriate.

We could:
1. Add `Empty` for nodes like `ASTEmptyReferenceExpression`, but allow nodes like `ASTField` to extend `AbstractApexNode` directly.
2. Find a name instead of `Empty` that is appropriate for both cases. (I am not sure what that would be.)

aaronhurst-google replied:
Right, the suggestion was that "special cases" (like `ASTField`) should inherit from `AbstractApexNode` directly and each implement (e.g.) `ASTField.calculateLineNumbers` as appropriate, but that the base class methods (e.g.) `AbstractApexNode.calculateLineNumbers` should remain abstract and unimplemented. These default implementations (like you have here, that make sense for "empty" methods) should be in a separate class `Empty` that would only be extended by nodes where it applies (like `ASTEmptyReferenceExpression`).

eklimo replied:
Please see [`bb813f9`](https://github.com/eklimo/pmd/pull/8/commits/bb813f9701fdbb9a85f00a7db59409cc937e5b95)

aaronhurst-google replied:
LGTM

aaronhurst-google commented on pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/ast/ASTLiteralExpression.java:
```diff
@@ -25,64 +26,71 @@ public Object jjtAccept(ApexParserVisitor visitor, Object data) {
-        if (node.getLiteral() != null) {
-            return String.valueOf(node.getLiteral());
+        if (isString()) {
+            return node.asCodeString().replace("'", "");
```
Let's just access the `LiteralExpression.StringVal.value` for the raw string value of a literal. (I think there are some other complexities of printing a literal "as a code string" that still need to be handled which will complexity this. For example, escaping single quotes inside the literal.)

aaronhurst-google commented on pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/ast/ASTMethodCallExpression.java:
```diff
@@ -22,9 +23,7 @@ public Object jjtAccept(ApexParserVisitor visitor, Object data) {
-        // return node.getMethodName();
-        // TODO(b/239648780)
-        return null;
+        return node.getId().asCodeString();
```
nit: prefer `node.getId().string` when you want Identifier name strings.

The `asCodeString()` helpers were intended for outputting AST subtrees back out as a code. They're obviously the same for an `Identifier`, but I I suspect the callers intention isn't to render as parseable Apex code.

eklimo replied:
I was debating this and decided on `asCodeString` to be consistent with other nodes that are `PrintableAsCodeString`. I have no preference though, so I will change the `Identifier` cases to use `string`.

By that logic, shouldn't the `PrintableAsCodeString` code I just implemented for literals also not be used in `ASTLiteralExpression.getImage`?

aaronhurst-google replied:
Perhaps. I'm less certain about the meaning of the `getImage` String for numeric literals to be able to offer much guidance. I defer to your judgment.

eklimo replied:
Discussed externally; resolved for now

zorzella commented on pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/ast/ASTLiteralExpression.java:
```diff
@@ -25,64 +26,71 @@ public Object jjtAccept(ApexParserVisitor visitor, Object data) {
+            return LiteralType.DOUBLE;
+        } else if (node instanceof LiteralExpression.LongVal) {
+            return LiteralType.LONG;
+        } else if (false) { // TODO(b/239648780)
```
just curious as to why this is the only not-implemented (presumably "to be implemented later"), since I would expect that LONG is just about identical to INTEGER...

eklimo replied:
It is `DECIMAL` that is unimplemented, not `LONG`. Summit currently has no concept of a `Decimal` literal.

zorzella commented on pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/ast/ApexTreeBuilder.kt:
```diff
@@ -142,6 +177,174 @@ class ApexTreeBuilder(val sourceCode: String, val parserOptions: ApexParserOptio
+            BinaryExpression.Operator.ALTERNATIVE_NOT_EQUAL,
+            BinaryExpression.Operator.EXACTLY_EQUAL,
+            BinaryExpression.Operator.EXACTLY_NOT_EQUAL,
+            BinaryExpression.Operator.BITWISE_XOR,
```
is a bitwise XOR a boolean expression? I would have expected this together with bitwise or; and here, I would have expected a logical xor...

eklimo replied:
That is a mistake, thank you for pointing this out

zorzella commented on pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/ast/ApexTreeBuilder.kt:
```diff
@@ -142,6 +177,174 @@ class ApexTreeBuilder(val sourceCode: String, val parserOptions: ApexParserOptio
+    ): Triple<List<Identifier>, Expression?, Boolean> =
+        when (node) {
+            is FieldExpression ->
+                if (node.isSafe) {
```
Is there a reasonable way to write a helper function called "flattenIfSafe" that can be called by all `is` branches here?

eklimo replied:
Resolved: N/A after [3fb57fa](https://github.com/eklimo/pmd/pull/8/commits/3fb57fa9d4ec31c55b1481417f53b0e179ec66e2)

eklimo committed:
Implement `AbstractApexNode.getDefiningType`

eklimo committed:
Fix `ASTLiteralExpression.getImage`

eklimo committed:
Implement reference types

- Implement `ASTReferenceExpression.isSObjectType`
- Implement `ASTReferenceExpression.hasRealLoc`

aaronhurst-google commented on pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/ast/AbstractApexNode.java:
```diff
@@ -128,11 +128,10 @@ public String getLocation() {
-        //     return definingType.getApexName();
-        // }
-        // TODO(b/239648780)
+        ApexRootNode<?> rootNode = this instanceof ApexRootNode ? (ApexRootNode<?>) this : getFirstParentOfType(ApexRootNode.class);
```
FYI I would avoid using `ApexRootNode` where possible. This interface went away in 7.x. It's equivalent to a node with a `null` parent, right?

eklimo replied:
It seems like [`ApexRootNode`](https://github.com/pmd/pmd/blob/master/pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/ast/ApexRootNode.java) was just replaced with a class named [`BaseApexClass`](https://github.com/pmd/pmd/blob/pmd/7.0.x/pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/ast/BaseApexClass.java). The subclasses (`ASTUserClass`, `ASTUserInterface`, etc.) all still extend it.

aaronhurst-google replied:
Thanks for that observation. In that case, the rebase onto 7.x seems straightforward enough, as-is.

aaronhurst-google reviewed (changes requested):
A few comments are outstanding.

eklimo committed:
Fix `ASTEmptyReferenceExpression.getDefiningType`

eklimo committed:
Implement `ASTMethodCallExpression.getFullMethodName`

eklimo committed:
Update `ASTLiteralExpression.getImage`

eklimo committed:
Replace `Triple` with `data class`

eklimo committed:
Fix `BITWISE_XOR`

eklimo committed:
Replace `Identifier.asCodeString` with `getString`

eklimo commented on pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/ast/ASTMethodCallExpression.java:
```diff
@@ -4,17 +4,24 @@
     @Deprecated
     @InternalApi
-    public ASTMethodCallExpression(CallExpression callExpression) {
+    public ASTMethodCallExpression(CallExpression callExpression, List<Identifier> components) {
```
Thoughts on passing additional data to wrapper nodes through the constructor? Previously, all methods could be determined based on only the properties of the internal node or the structure of the AST. Do these additional parameters make the design too dependent on the tree builder design (i.e. if the tree builder is restructured, the implementation here may have to change)?

aaronhurst-google replied:
I think it's a fine solution, particularly if the additional data comes from a sibling in the Summit tree structure.

In general, another strategy would be to construct from a single node that is higher in the Summit tree, such that all properties could be computed from its subtree. But if that higher node is too unrelated to a single CallExpression it might make less sense.

eklimo committed:
Fix call expressions

- Only the `receiver` of `CallExpression`s should be flattened

aaronhurst-google commented on pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/ast/ASTMethodCallExpression.java:
```diff
@@ -4,16 +4,27 @@
+     * The {@link Identifier}s that constitute the {@link CallExpression#getReceiver() receiver} of
+     * this method call.
+     */
+    private final List<Identifier> components;
```
suggesting naming the variable `reciever` or `recieverComponents`. (Also the ctor parameter)

eklimo committed:
Fix `referenceTypeOf`

Reference type should only be `STORE` when the variable is the target of
assignment.

eklimo committed:
Rename `ASTMethodCallExpression.components`

eklimo committed:
Refactor `AbstractApexNode`
